channeldb/payments.go

```diff
@@ -252,9 +252,74 @@ type Payment struct {
 	Failure *FailureReason
 }
 
+// ToMPPayment converts a legacy payment into an MPPayment.
+func (p *Payment) ToMPPayment() *MPPayment {
+	var (
+		htlcs   []HTLCAttempt
+		settle  *HTLCSettleInfo
+		failure *HTLCFailInfo
+	)
+
+	// Promote the payment preimage to proper settle struct, if it exists.
+	if p.Preimage != nil {
+		// NOTE: SettleTime is not set for legacy payments.
+		settle = &HTLCSettleInfo{
+			Preimage: *p.Preimage,
+		}
+	}
+
+	// Promote the payment failure to a proper fail struct, if it exists.
+	if p.Failure != nil {
+		// NOTE: FailTime is not set for legacy payments.
+		failure = &HTLCFailInfo{
+			Reason: *p.Failure,
+		}
+	}
+
+	// Either a settle or a failure may be set, but not both.
+	if settle != nil && failure != nil {
+		panic("htlc attempt has both settle and failure info")
+	}
+
+	// Populate a single HTLC on the MPPayment if an attempt exists on the
+	// legacy payment. If none exists we will leave the attempt info blank
+	// since we cannot recover it.
+	if p.Attempt != nil {
+		// NOTE: AttemptTime is not set for legacy payments.
+		htlcs = []HTLCAttempt{
+			{
+				PaymentID:  p.Attempt.PaymentID,
+				SessionKey: p.Attempt.SessionKey,
+				Route:      p.Attempt.Route,
+				Settle:     settle,
+				Failure:    failure,
+			},
+		}
+	} else {
+		htlcs = []HTLCAttempt{
```

Comment:
Isn't it a possibility that the payment was never even attempted and that `p.Attempt` is therefore `nil`? In that case it doesn't seem right to add an `HTLCAttempt`

Reply:
I prefer `HtlcAttempt` because it doesn't stand out so much. To me "Htlc" is a word :)

Reply:
good catch, also modified the test to support `Fail` after registering so that we could test the case where the failure and attempt are both populated.